docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -0,0 +1,210 @@
+
+To achieve this, ClusterResourceSet CRD is introduced that will be responsible for applying a set resources defined by users to the matching clusters.
+
+### Goals
+
+- Provide a means to specify a set of resources to apply automatically to newly-created and existing Clusters. Resources initially will be applied only once; full sync functionality could be added later.
+- Support additions to the resource list by applying the new added resources to both new and existing matching clusters.
+- Support both json and yaml resources.
+
+### Non-Goals/Future Work
+
+- Replace or compete with the Cluster Addons subproject
+- Support deletion of resources from clusters. Deleting a resource from a ClusterResourceSet or deleting a ClusterResourceSet does not result in deletion of those resources from clusters.
+- Lifecycle management of the installed resources (such as CNI)
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As someone creating multiple clusters, I want all my clusters to have a CNI provider of my choosing installed automatically, so I don’t have to manually repeat the installation for each new cluster.
+
+#### Story 2
+
+As someone creating multiple clusters, I want all my clusters to have a StorageClass installed automatically, so I don't have to manually repeat the installation for each new cluster.
+
+#### Story 3
+
+As someone creating multiple clusters, I want to be able to provide different values for some fields in the resources for different clusters. For example, CNIs podCIDRs  may be required to be distinct for each cluster, hence some templating mechanism for variable substitution in the resources is needed.
+
+### Implementation Details/Notes/Constraints
+
+#### Data model changes
+
+None. We are planning to implement this feature without modifying any of the existing structure to minimize the footprint of ClusterResourceSet Controller. This enhancement will follow Kubernetes’s feature-gate structure and will be under the experimental package with its APIs, and enabled/disabled with a feature gate. 
+
+#### ClusterResourceSet Object Definition
+
+This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.
+
+The resources field is a list of secrets/configmaps in the same namespace. The clusterSelector field is a Kubernetes [label selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) that matches against labels on clusters (in any namespace).
+
+*Sample ClusterResourceSet YAML*
+
+```yaml
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: ClusterResourceSet
+metadata:
+ name: postcreate-conf
+ namespace: default
+spec:
+ mode: "ApplyOnce"
+ clusterSelector:
```

Comment:
What about adding an optional namespace field in cluster selector, so a ClusterResourceSet can be easily scoped to clusters in a single namespace

Reply:
I posted in an `outdated` thread a similar question. If we can align with the CAPA multi-tenancy proposal, then we'll have a similar UX.

https://github.com/kubernetes-sigs/cluster-api/pull/3050#discussion_r425016811

Reply:
Combining with the thread immediately above this one and pulling in Naadir's outdated comment, I think this is what's being proposed (names TBD, of course):

**GlobalClusterResourceSet**
- Cluster-scoped, so you need "higher"/specific RBAC permissions to be able to create these
- `spec.namespaceSelector` governs which namespaces are covered by this GlobalClusterResourceSet
- `spec.clusterSelector` has the same semantics as before
- If you create a Cluster in a covered namespace, and it's covered by `clusterSelector`, apply the resources
- If you create a Cluster in some other, non-covered namespace, it is ignored

**ClusterResourceSet**
- Namespace-scoped
- `spec.clusterSelector` only matches Clusters in the same namespace

Even with this split, I suppose you could have permissions to create a ClusterResourceSet in a namespace but not to get Secrets and/or ConfigMaps in the namespace, and if you then create a Cluster, you could cause CAPI to apply the Secrets/ConfigMaps into your new Cluster. I'm not sure how likely this specific set of permissions would be, though.

Reply:
+1 for the split

Reply:
Changed ClusterResourceSet to be only applied to the clusters in the same namespace which will solve security concerns.